eras/babbage/impl/src/Cardano/Ledger/Babbage/PParams.hs

```diff
@@ -0,0 +1,533 @@
+      _maxBHSize = fromSMaybe (_maxBHSize pp) (_maxBHSize ppup),
+      _keyDeposit = fromSMaybe (_keyDeposit pp) (_keyDeposit ppup),
+      _poolDeposit = fromSMaybe (_poolDeposit pp) (_poolDeposit ppup),
+      _eMax = fromSMaybe (_eMax pp) (_eMax ppup),
+      _nOpt = fromSMaybe (_nOpt pp) (_nOpt ppup),
+      _a0 = fromSMaybe (_a0 pp) (_a0 ppup),
+      _rho = fromSMaybe (_rho pp) (_rho ppup),
+      _tau = fromSMaybe (_tau pp) (_tau ppup),
+      _protocolVersion = fromSMaybe (_protocolVersion pp) (_protocolVersion ppup),
+      _minPoolCost = fromSMaybe (_minPoolCost pp) (_minPoolCost ppup),
+      _coinsPerUTxOWord = fromSMaybe (_coinsPerUTxOWord pp) (_coinsPerUTxOWord ppup),
+      _costmdls = fromSMaybe (_costmdls pp) (_costmdls ppup),
+      _prices = fromSMaybe (_prices pp) (_prices ppup),
+      _maxTxExUnits = fromSMaybe (_maxTxExUnits pp) (_maxTxExUnits ppup),
+      _maxBlockExUnits = fromSMaybe (_maxBlockExUnits pp) (_maxBlockExUnits ppup),
+      _maxValSize = fromSMaybe (_maxValSize pp) (_maxValSize ppup),
+      _collateralPercentage = fromSMaybe (_collateralPercentage pp) (_collateralPercentage ppup),
+      _maxCollateralInputs = fromSMaybe (_maxCollateralInputs pp) (_maxCollateralInputs ppup)
+    }
+
+-- ===================================================
+-- Figure 1: "Definitions Used in Protocol Parameters"
+
+-- The LangDepView is a key value pair. The key is the (canonically) encoded
+-- language tag and the value is the (canonically) encoded set of relevant
+-- protocol parameters
+data LangDepView = LangDepView {tag :: ByteString, params :: ByteString}
+  deriving (Eq, Show, Ord, Generic, NoThunks)
+
+getLanguageView ::
+  forall era.
+  PParams era ->
+  Language ->
+  LangDepView
+getLanguageView pp lang@PlutusV1 =
+  LangDepView -- The silly double bagging is to keep compatibility with a past bug
+    (serialize' (serialize' lang))
+    ( serialize'
+        ( serializeEncoding' $
+            maybe encodeNull enc $
+              Map.lookup lang (_costmdls pp)
+        )
+    )
+  where
+    enc (CostModel cm) = encodeFoldableAsIndefinite $ Map.elems cm
+getLanguageView pp lang@PlutusV2 =
+  LangDepView
+    (serialize' lang)
+    (serializeEncoding' $ maybe encodeNull toCBOR $ Map.lookup lang (_costmdls pp))
+
+encodeLangViews :: Set LangDepView -> Encoding
+encodeLangViews views = encodeMapLen n <> foldMap encPair ascending
+  where
+    n = fromIntegral (Set.size views) :: Word
+    ascending = sortBy (shortLex `on` tag) $ Set.toList views
+    encPair (LangDepView k v) = encodePreEncoded k <> encodePreEncoded v
+    shortLex :: ByteString -> ByteString -> Ordering
+    shortLex a b
+      | BS.length a < BS.length b = LT
+      | BS.length a > BS.length b = GT
+      | otherwise = compare a b
+
+-- | Turn an PParams' into a Shelley.Params'
+retractPP :: HKD f Coin -> HKD f UnitInterval -> HKD f Nonce -> PParams' f era2 -> Shelley.PParams' f era1
```

Comment:
ditto

Reply:
this function is already as meaningfully abstract as it can be i think

Reply:
oh yes, you can use the Alonzo one because of `d` and `extraEntropy`, makes sense now!